README
=== pm-utils ===

What is pm-utils?

* Pm-utils provides simple shell command line tools to suspend and
  hibernate computer that can be used to run vendor, distribution, or
  user supplied scripts on suspend and resume.

Why do we need it?

* All the main distributions  are re-implementing the same
  functionality to support common power management
  operations. pm-utils provides a standardized distribution agnostic
  layout and interface for common power management tasks, which frees
  software and hardware vendors from having to maintain
  distribution-specific suspend/resume functionality.

* Suspend/resume functionality can be easily modified by installing
  files into the /etc/pm/sleep.d directory.  These files, known as
  "hooks" can perform specific tasks on suspend/resume, such as:
   
   - enabling and disabling standby LED's on laptop hardware 
   - enabling suspend GUI's like suspend2 
   - re-enabling video
   - starting and stopping services that can't cope with suspending 
   - re-syncing the time with ntp 
   - removing and modprobing modules when needed 
   - setting grub to be the default target for a hibernate-resume 
   - other wacky things that need doing on specific systems

How do "hooks" work?

* You put an executable file in /etc/pm/sleep.d.  When suspend or
  hibernate is called, several things happen:
  
  1) The files in /etc/pm/config.d/* are evaluated in C sort order.  
     These files can be provided by individual packages outside of
     pm-utils, and contain globally available configuration settings
     for pm-utils and the hooks.
  2) Each of the hooks in /etc/pm/sleep.d/* are executed in C sort
     order, with a command line argument of either "suspend" or "hibernate". 
  3) The system suspends or hibernates.
  4) Some event happens to wake the machine up
  5) Each of /etc/pm/sleep.d/* are executed in reverse C sort order,
     with a command line argument of "resume" or "thaw".

End-user customization and debugging:

* If a particular hook is causing problems on your system, you can disable it
  using the HOOK_BLACKLIST environment variable by creating a file in 
  /etc/pm/config.d and adding the line:
  HOOK_BLACKLIST="hookname 99another-hook"

* If a parameter (or lack thereof) passed to pm-suspend and friends is causing
  problems, or you need to debug the suspend/resume process to work out what 
  quirks are causing problems, you can use the ADD_PARAMETERS and 
  DROP_PARAMETERS environment variables.  To do this, create a file in 
  /etc/pm/config.d, and add the following lines to it:
  ADD_PARAMETERS="--parameter-to-add --another-parameter"
  DROP_PARAMETERS="--parameter-to-drop --another-parameter-to-drop"

  If you want to drop all parameters (for testing purposes, or to work around 
  bugs in HAL), you can use DROP_PARAMETERS="all"

* If you suspect that a kernel module is preventing you from being
  able to suspend and resume, you can use the SUSPEND_MODULES
  environment variable to have that module removed when the system
  suspends and reloaded when the system wakes up.

* To find out what parameters can be passed to pm-suspend and friends, run them
  with '--help' as the first parameter as root.  This will print out the 
  options that it supports and which hooks or modules handle those options.

  That's it!



NEWS
Major changes since the 1.3 series:

* The pm-utils git tree includes infrastructure to build .deb packages,
  .rpm packages. and Arch Linux .tar packages.  These are not intended
  to replace distro packaging, and are not included in the release tarballs.

* We include some basic power management hooks. The individual hooks can
  conflict with the ones provided by the laptop-mode-tools scripts, so you
  should either just remove laptop-mode-tools entirely or pick and choose
  between which package should handle what powersaving functionality.

Major changes since the 1.2 series:
 
* We no longer rely on HAL at all for video quirks. 

* The pm-utils project has taken over maintenance of the video quirk database.
  Most systems coming out no longer rely on quirks, due to KMS enabled drivers
  for most common chipsets along with the two major binary video drivers
  handling any needed suspend/resume handling themselves.

Major changes since the 1.1 series:

* Our pkg-config support now includes variables that directly map to where
  the various hooks should be installed.  This should make it easier for
  other packages to install hooks in a portable fashion.

* 90clock no longer runs by default.  It added over a second to suspend,
  and the vast majority of hardware does not need it to keep the clocks
  in sync.  If a user needs this hook, they can set the NEED_CLOCK_SYNC
  environment variable.

  If pm-utils@lists.freedesktop.org does not get any reports of systems
  that require this hook to keep their clocks synced, it will be removed
  in a future release.

* 50ntpd has been removed.  It was very ad-hoc to begin with, and there
  has never been a guarantee that restarting ntpd would force it to resync
  the clocks, or that the system would even find a network after coming out
  of suspend.  The ntpd daemon itself is robust in the face of interfaces
  going away and reappearing, so there is no reason to restart it on account
  of the network interfaces changing -- ntpd will keep the clock updated over
  any available interface according to its own schedule.

  If you just want to update the time whenever the network comes back without
  running ntpd continually, utilize NetworkManager dispatch scripts (or your
  network infrastructure equivalent) to do so whenever the Internet comes up.

* 65alsa has been removed. According to the ALSA dev team, all drivers that
  have power management support will save and restore their state correctly
  across a suspend or hibernate, and drivers that don't have power management
  support need to have their state saved, then unloaded before suspend, 
  reloaded on wakeup, and then have their state restored.

  The best way to take care of this would be to use the standard pm-utils 
  mechanisms for loading and unloading modules, and then arrange for modprobe
  and/or udev to save and restore alsa state when unloading or loading alsa
  modules.

* 55battery has been removed.  This hook was in place to work around a HAL
  issue with rescanning battery status on resume, but HAL implemented a fix
  for this issue in 2006.  

* We always set the acpi video flags to the values indicated by HAL.  If HAL
  does not indicate that a specific set of acpi video flags are needed, we
  always set it to 0.  As a consequence, the acpi video flags passed to the
  kernel at boot time are ignored. If you rely on this functionality and HAL
  does not have the appropriate quirks for your system, we strongly recommend
  that you create a .fdi file that has the appropriate quirks for your system
  by passing the appropriate quirks to pm-suspend along with the 
  --save-quirks-as-fdi parameter.  This will create a .fdi file that you can
  then use locally (after restarting HAL) and submit for inclusion to hal-info
  so that other users with your hardware config can benifit from your actions.

* Sleep backends are now stackable, and pm-utils will always fall back to
  kernel methods if nothing else will work on any given system.
  See pm/HOWTO.modules for more information.

* pm-utils now knows how to get the quirks from HAL, and includes functionality
  to save user-passed quirks in a .fdi file.  This makes it easier for users
  to generate and share quirks, and it makes debugging suspend/resume problems
  easier.

* We now have a recommended hook ordering convention.  See pm/HOWTO.hooks for
  more information.  All pm-utils supplied hooks have been renamed to conform
  to this hook ordering.




src/pm-pmu.c
/*
 * suspend functions for machines with Mac-style pmu
 *
 * Copyright 2006 Red Hat, Inc.
 *
 * Based on work from:
 *    Peter Jones <pjones@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#define _GNU_SOURCE 1

#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdint.h>

/* XXX this is a total hack for bad system headers. */
typedef u_int32_t __u32;
#include <linux/pmu.h>

#if !defined(PMU_IOC_CAN_SLEEP) || !defined(PMU_IOC_SLEEP)
#error you must have working pmu kernel headers installed to build this program.
#endif

static int
pmu_can_sleep(const int fd)
{
    unsigned long arg = 0;
    if (ioctl(fd, PMU_IOC_CAN_SLEEP, &arg) < 0 || arg != 1)
        return 1;
    return 0;
}
    
static int
pmu_sleep(const int fd)
{
    unsigned long arg = 0;

    if (ioctl(fd, PMU_IOC_SLEEP, arg) < 0)
        return 1;
    return 0;
}

static inline int
print_usage(FILE *output, int retval)
{
    fprintf(output, "usage: pm-pmu --suspend | --check\n");
    return retval;
}

int main(int argc, char *argv[])
{
    int task=0,fd=0,ret=0;
    if (argc != 2)
        return print_usage(stderr, 1);
    if (!strcmp(argv[1], "--help")) {
        task=1;
    } else if (!strcmp(argv[1],"--check")) {
        task=2;
    } else if (!strcmp(argv[1],"--suspend")) {
        task=3;
    }
    switch (task) {
        case 0: ret = print_usage(stderr, 1); break;
        case 1: ret = print_usage(stdout, 0); break;
        case 2: /* this is intentional */
        case 3:
            if (access("/dev/pmu", W_OK)) {
                ret = 1;
            } else {
                if ((fd = open("/dev/pmu", O_RDWR)) < 0) {
                    perror("open");
                    return 1;
                }
            
                ret = pmu_can_sleep(fd);
                if (!ret && task == 3) 
                    ret = pmu_sleep(fd);
                close(fd);
            }
            break;
    }
   return ret; 
}

/*
 * vim:ts=8:sw=4:sts=4:et
 */


on_ac_power.c
#include <libhal.h>

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
	LibHalContext *ctx;
	char **devs;
	DBusError error;
	DBusConnection *conn;
	int num, x, rc = 255;
	
	ctx = libhal_ctx_new();
	if (!ctx) {
		return 255;
	}
	dbus_error_init(&error);
	conn = dbus_bus_get (DBUS_BUS_SYSTEM, &error);
	if (!conn) {
		return 255;
	}
	if (!libhal_ctx_set_dbus_connection(ctx, conn)) {
		return 255;
	}
	if (!libhal_ctx_init(ctx,&error)) {
		return 255;
	}
	/* Check for AC/DC/etc adapters */
	devs = libhal_find_device_by_capability(ctx, "ac_adapter", &num, &error);;
	if (num) {
		for (x = 0; x < num ; x++) {
			dbus_bool_t is_present;
			
			is_present = libhal_device_get_property_bool(ctx, devs[x], "ac_adapter.present", &error);
			if (!dbus_error_is_set(&error) && is_present) {
				rc = 0;
				goto out;
			}
		}
	}
	/* If there are none, check for batteries. If there are no
	   batteries either, assume AC. */
	devs = libhal_find_device_by_capability(ctx, "battery", &num, &error);;
	if (num) {
		rc = 1;
	} else {
		rc = 0;
	}
out:
	libhal_ctx_shutdown (ctx, &error);
	libhal_ctx_free (ctx);
	dbus_connection_disconnect (conn);
	return rc;
}


src/pm-reset-swap.c
/*
 * mark a swap device as not usable for thaw
 *
 * Copyright 2006 Red Hat, Inc.
 *
 * Authors:
 *   Peter Jones <pjones@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#define _GNU_SOURCE 1

#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdint.h>

/* XXX this is a total hack for bad system headers. */
typedef u_int32_t __u32;
#include <linux/pmu.h>

typedef u_int64_t sector_t;

/*
 * Here's what the disk looks like after a successful hibernate:

 00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 *
 00000400  01 00 00 00 ff df 03 00  00 00 00 00 00 00 00 00  |................|
 00000410  00 00 00 00 00 00 00 00  00 00 00 00 73 77 61 70  |............swap|
 00000420  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 *
 00000fe0  00 00 00 00 01 00 00 00  00 00 00 00 53 57 41 50  |............SWAP|
 00000ff0  53 50 41 43 45 32 53 31  53 55 53 50 45 4e 44 00  |SPACE2S1SUSPEND.|

 * And here's what it looks like if you haven't hibernated to it:

 00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 *
 00000400  01 00 00 00 ff df 03 00  00 00 00 00 00 00 00 00  |................|
 00000410  00 00 00 00 00 00 00 00  00 00 00 00 73 77 61 70  |............swap|
 00000420  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 *
 00000ff0  00 00 00 00 00 00 53 57  41 50 53 50 41 43 45 32  |......SWAPSPACE2|

 *
 */

/* The signature offset is annoying.  The struct is:
 * struct swsusp_header {
 *     char reserved[PAGE_SIZE - 20 - sizeof(sector_t)];
 *     sector_t image;
 *     char orig_sig[10];
 *     char sig[10];
 * }
 */
off_t get_sig_offset(int orig)
{
    int pagesize = sysconf(_SC_PAGESIZE);

    return (off_t)((sizeof (char) * pagesize) - (orig ? 20 : 10));
}

static inline int
print_usage(FILE *output, int retval)
{
    fprintf(output, "usage: pm-reset-swap <device node>\n");
    return retval;
}

int check_resume_block(FILE *dev, off_t offset)
{
    char buf[11] = {0,};
    off_t location = offset + get_sig_offset(0);

    if (fseek(dev, location, SEEK_SET) < 0)
        return -1;
    if (fread(buf, sizeof (char), 10, dev) != 10)
        return -1;

    if (!strncmp(buf, "S1SUSPEND", 9) ||
            !strncmp(buf, "ULSUSPEND", 9))
        return 1;

    return 0;
}

int clear_resume_block(FILE *dev, off_t offset)
{
    char buf[21] = {0,};
    off_t location = offset + get_sig_offset(1);

    if (fseek(dev, location, SEEK_SET) < 0)
        return -1;

    if (fread(buf, sizeof (char), 20, dev) != 20)
        return -1;

    if (fseek(dev, location, SEEK_SET) < 0)
        return -1;

    memmove(buf+10, buf, 10);
    memset(buf, '\0', 10);

    if (fwrite(buf, sizeof (char), 20, dev) != 20)
        return -1;

    return 0;
}

int main(int argc, char *argv[])
{
    FILE *dev = NULL;
    int rc;

    if (argc != 2)
        return print_usage(stderr, 1);

    if (!strcmp(argv[1], "--help")) {
        return print_usage(stdout, 0);
    }

    if (!(dev = fopen(argv[1], "r+b"))) {
        fprintf(stderr, "Could not open \"%s\": %m\n", argv[1]);
        return 1;
    }

    rc = check_resume_block(dev, 0);
    if (rc < 0) {
        fprintf(stderr, "Could not check \"%s\" for swap signature: %m\n",
                argv[1]);
        fclose(dev);
        return 2;
    }

    if (rc == 1) {
        if (clear_resume_block(dev, 0)) {
            fprintf(stderr, "Could not clear swap signature on \"%s\": %m\n",
                    argv[1]);
            fclose(dev);
            return 1;
        }
    }
        
    fclose(dev);
    return 0;
}

/*
 * vim:ts=8:sw=4:sts=4:et
 */
